src/lib/transition_frontier_persistence/transition_frontier_persistence.ml

```diff
@@ -106,19 +53,180 @@ end = struct
       (Transition_storage.get transition_storage ~key:state_hash)
       ~f:to_verified
 
-  let remove_transitions {transition_storage; _} state_hashes =
-    Transition_storage.set_batch transition_storage ~remove_keys:state_hashes
-      ~update_pairs:[]
+  let read_root {root_storage; _} =
+    Deferred.Result.map_error
+      (Root_storage.load root_storage.controller root_storage.path)
+      ~f:(function
+      | `Checksum_no_match -> Error.of_string "Checksum did not match"
+      | `IO_error e -> Error.createf !"IO_error: %s" (Error.to_string_hum e)
+      | `No_exist -> Error.createf !"File %s does not exist" root_storage.path )
 
   let lift_or_error deferred_x =
     let%map x = deferred_x in
     Or_error.return x
 
+  module Batch = struct
+    let set_transition batch
+        {With_hash.hash= state_hash; data= external_transition} =
+      Transition_storage.Batch.set batch ~key:state_hash
+        ~data:(External_transition.of_verified external_transition)
+
+    let remove_transition batch state_hash =
+      Transition_storage.Batch.remove batch ~key:state_hash
+  end
+
   let parent_hash transition =
     let open External_transition.Verified in
     let protocol_state = protocol_state transition in
     External_transition.Protocol_state.(previous_state_hash protocol_state)
 
+  module Frontier_diff = struct
+    type add_transition =
+      (External_transition.Verified.t, State_hash.t) With_hash.t
+
+    type move_root =
+      { best_tip: (External_transition.Verified.t, State_hash.t) With_hash.t
+      ; removed_transitions: State_hash.Stable.Latest.t list
+      ; new_root: State_hash.Stable.Latest.t
+      ; new_scan_state: Staged_ledger.Scan_state.Stable.Latest.t }
```

Comment:
I would still recommend removing `removed_transitions` from this diff and instead having a deterministic order for removed transitions and model them as part of the mutant of `Move_root` diffs.